bulkcopy.go

```diff
@@ -0,0 +1,576 @@
+	}
+
+	if b.Debug {
+		log.Println(query)
+	}
+	_, err = stmt.Exec(nil)
+	if err != nil {
+		return err
+	}
+
+	b.headerSent = true
+
+	var buf = b.cn.sess.buf
+	buf.BeginPacket(packBulkLoadBCP)
+
+	//columns
+	columnMetadata := b.createColMetadata()
+	buf.Write(columnMetadata)
+
+	return nil
+}
+
+func (b *MssqlBulk) AddRow(row []interface{}) (err error) {
+
+	if b.headerSent == false {
+		err = b.sendBulkCommand()
+		if err != nil {
+			return
+		}
+	}
+
+	if len(row) != len(b.bulkColumns) {
+		return fmt.Errorf("Row does not have the same number of columns than the destination table %d %d", len(row), len(b.bulkColumns))
+	}
+
+	bytes, err := b.makeRowData(row)
+	b.cn.sess.buf.Write(bytes)
+
+	b.numRows = b.numRows + 1
+	return err
+}
+
+func (b *MssqlBulk) AddRows(rows [][]interface{}) (err error) {
+	for _, row := range rows {
+		err = b.AddRow(row)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (b *MssqlBulk) makeRowData(row []interface{}) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(tokenRow)
+
+	logcol := ""
+	for i, col := range b.bulkColumns {
+
+		if b.Debug {
+			logcol = logcol + fmt.Sprintf(" col[%d]='%v' ", i, row[i])
+		}
+		param, err := b.makeParam(row[i], col)
+		if err != nil {
+			return nil, fmt.Errorf("bulkcopy: %s", err.Error())
+		}
+
+		if col.ti.Writer == nil {
+			log.Fatalf("no writer for column: %s, TypeId: %#x", col.ColName, col.ti.TypeId)
+		}
+		err = col.ti.Writer(buf, param.ti, param.buffer)
+		if err != nil {
+			return nil, fmt.Errorf("bulkcopy: %s", err.Error())
+		}
+	}
+	if b.Debug {
+		log.Printf("row[%d] %s\n", b.numRows, logcol)
+	}
+
+	return buf.Bytes(), nil
+
+}
+
+func (b *MssqlBulk) Done() (rowcount int64, err error) {
+	if b.headerSent == false {
+		//no rows had been sent
+		return 0, nil
+	}
+	var buf = b.cn.sess.buf
+	buf.WriteByte(tokenDone)
+
+	binary.Write(buf, binary.LittleEndian, uint16(doneFinal))
+	binary.Write(buf, binary.LittleEndian, uint16(0)) //     curcmd
+
+	if b.cn.sess.loginAck.TDSVersion >= verTDS72 {
+		binary.Write(buf, binary.LittleEndian, uint64(0)) //rowcount 0
+	} else {
+		binary.Write(buf, binary.LittleEndian, uint32(0)) //rowcount 0
+	}
+
+	buf.FinishPacket()
+
+	//
+	tokchan := make(chan tokenStruct, 5)
+	go processResponse(context.Background(), b.cn.sess, tokchan)
+
+	var rowCount int64
+	for token := range tokchan {
+		switch token := token.(type) {
+		case doneStruct:
+			if token.Status&doneCount != 0 {
+				rowCount = int64(token.RowCount)
+			}
+		case error:
+			return 0, token
+		}
+	}
+	return rowCount, nil
+}
+
+func (b *MssqlBulk) createColMetadata() []byte {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(tokenColMetadata)                                    // token
+	binary.Write(buf, binary.LittleEndian, uint16(len(b.bulkColumns))) // column count
+
+	for i, col := range b.bulkColumns {
+
+		if b.cn.sess.loginAck.TDSVersion >= verTDS72 {
+			binary.Write(buf, binary.LittleEndian, uint32(col.UserType)) //  usertype, always 0?
```

Comment:
Thanks for the review. There's a difference, it is writing 4 bytes instead of 2 bytes.